### PR TITLE
MOOSE Training and New Workshop - Intro, Tutorial Header, and Step 1-5

### PR DESCRIPTION
**Pull request by crswong888**
# Updated Description (7/30/20)

This PR reorganizes the "Getting Started" section of the website to include all examples and all future tutorials as part of this procedure intended for new MOOSE users. The idea is that the reader proceed in the following order: "Install MOOSE" -> "New Users" -> "Examples and Tutorials."

This PR moves the `examples/` directory under `getting_strarted/examples_and_tutorials/examples/`, and moves its index content to `getting_started/examples_and_tutorials/index.md`. The header page, "Examples and Tutorials," includes an introduction, an "Examples" section, a "Tutorials" section, and a "Workshop" section to present the respective materials available for all three of these training types. A scrollbox MooseDocs feature was created under the Layout Extension to help format this page.

The "New Users" section includes everything from the old "Getting Started" page, except the install and update MOOSE instructions. I've included a section explaining how to enable physics modules in the applications `Makefile`. I've also added a section, "Helpful Software," to recommend that new users consider installing Atom, PEACOCK, and/or Paraview.

I've begun to develop a novel tutorial: "Tutorial 1: Application Development", which resembles the classic `darcy_thermo_mech` tutorial from the workshop. A link to this tutorial has been added to the "Examples and Tutorials" page. So far I have completed steps 1-5 of this tutorial. I created the necessary scripting in MooseDocs python to include content from `idaholab/babbler` needed for the tutorial demonstrations. 

I also created a new help page: "What is the Finite Element Method?" For which, I reference in the FAQ, the "Finite Element Concepts" page, and the tutorial. The discussion included serves the purpose of describing the core philosophy of the FEM, which is necessary for MOOSE developers to understand.

I needed to move `large_media/tutorials/` to `large_media/tutorials/darcy_thermo_mech/` so that I may distinguish media files associated with different tutorials, e.g., `large_media/tutorials/tutorial01_app_development/step02_result.png` and `large_media/tutorials/darcy_thermo_mech/step02_result.png`. To be clear, these two files are distinct - I actually cropped `step02_result.png` for Tutorial 1 so that it lays out more neatly on a MooseDown page. This change is reflected in my [`large_media:workshop-15232`](https://github.com/idaholab/large_media/tree/workshop-15232) branch. I updated all calls to the Media Extension in the workshop docs to be `!media darcy_thermo_mech/file` instead of `!media file`.

## Updated Description (6/30/20)

This PR initializes a novel approach to MOOSE new user training. The classic `darcy_thermo_mech` tutorial that is, typically, presented for MOOSE Workshops, shall be converted into a MooseDown format that appears much like the rest of the website. It shall be filed under a new section of the website: "Training", which will appear as a new mega menu. The current MOOSE Examples shall also appear within this new section. All future tutorials shall be appended to this portion of MOOSE.

The source code for tutorials and examples shall be located at `moose/training/tutorials` and `moose/training/examples`, respectively. Similarly, any media required for MooseDown content shall be filed under `large_media/training/tutorials` or `large_media/training/examples`. Finally, all new website content shall be filed under `modules/doc/training/*`. 

After most steps of the New Workshop tutorial, the reader will be asked to reproduce the problem themselves and commit their work to their own repositories. A new idaholab repository, `babbler`, was created to serve as the answer key for these demonstrations. It shall be added to MOOSE as a set of submodules - one for each commit, so that its entire history exists simultaneously and can be referenced as part of the Workshop content.

## Description (5/15/20)

Wrote the introductory potion of, as well as Step 1 of the `darcy_thermo_mech` tutorial for, the new MOOSE workshop - which converts the usual live workshops, supplemented with slides, into a virtual, MOOSEDown format that appears much like the rest of the website.

All of this is a major WIP, but this first merge will allow us to buff out some basic formatting issues before we start cranking out more pages. The final organization of the directories and their docs should be settled here.

(refs #15232 and #13617)

### CODE REVIEW

File: modules/doc/content/application_development/workshop/preface.md
Comment (aeslaughter) [inline modules/doc/content/application_development/workshop/preface.md:18]:
```suggestion
!equation id=stiffness
\bold{A} x = b
```

Comment (aeslaughter) [reply modules/doc/content/application_development/workshop/preface.md:18]: Please use this syntax for equations

**Review by aeslaughter — COMMENTED** [review]
This is a great start. I am going to push some re-organization changes and then provide some general comments.

**Comment by aeslaughter** [comment]
Please move the examples to training/examples, remove the links to examples and the workshop in the main Documentation tab, and then create a new "Examples" tab that will contain links to the pages you are working on.

File: modules/doc/content/training/darcy_thermo_mech/step01.md
Comment (aeslaughter) [inline modules/doc/content/training/darcy_thermo_mech/step01.md:1]: Can you rename this file to step01_moose_object.md. I just like the filename giving some information about what it contains.

**Comment by aeslaughter** [comment]
Another thought I have is that we should probably make the tutorial text complete from an application development standpoint. Step 01 should ask the user to create a github account, run stork, compile, run_tests, and push a commit. Then step02, create a problems directory, add a folder for diffusion, build input file, run it. Step 03, create a test for the input file you just made (syntax check in problems) and a smaller version in tests directory?

**Comment by crswong888** [comment]
> Another thought I have is that we should probably make the tutorial text complete from an application development standpoint. Step 01 should ask the user to create a github account, run stork, compile, run_tests, and push a commit. Then step02, create a problems directory, add a folder for diffusion, build input file, run it. Step 03, create a test for the input file you just made (syntax check in problems) and a smaller version in tests direc

**Comment by crswong888** [comment]
@aeslaughter I made the changes you requested and some of the other ones we talked about. Please take a moment to read through my Step 1 page so we can discuss how we want to integrate more hands-on stuff. In the meantime, I will look into the issue of expanding the contents extension to create a TOC with an outline style.

File: modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md
Comment (aeslaughter) [inline modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md:1]: The MooseObject stuff should come in step 2, before the first Kernel is created. This step should focus on Diffusion and the input file.

Comment (crswong888) [reply modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md:1]: The reason I included that first was because I though it related well to input file syntax. Perhaps I should only briefly mention them and then move this more in-depth discussion to step 2. Although, I wanted to talk about Galerkin FEM and weak forms in step 2, so that step would become cluttered. Let me know what you think of this.

Comment (aeslaughter) [reply modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md:1]: Step 2 can just be MooseObject; we can inherit from Diffusion originally so can leave all the FEM stuff for step3, which can use KernelGrad as a base class. This is just an idea, what do you think?

Comment (crswong888) [reply modules/doc/content/training/darcy_thermo_mech/step01_moose_object.md:1]: I like this idea, we should spread out concept discussions as much as possible. 

I'm thinking I subdivide FEM even further, and only talk about weighted residuals and generating a weak form in step 3. Then, maybe step 4 can be about polynomial fitting, test functions = shape functions in Galerkin FEM, jacobians, etc. Finally step 5 can be about integration, i.e., Gauss quad rules. If I have to create all new tutorial steps to make this work I will.

Anyways, first I'll work on step 2 with MOOSE objects, then we'll worry about FEM steps.

**Comment by crswong888** [comment]
@aeslaughter I've pretty much addressed your comments, and now I'm sort of just proceeding along with more workshop content. Things are starting to get messy, and I have unfinished things lurking everywhere, but its fine, I'm keeping track of this.

We need to consider if we want to shift the focus of this workshop to the user building their own application. I foresee an approach where we create an entirely new repo under idaholab account, and make each step a commit, or its own branch, but still require the reader make their own repo and reproduce ours. Let me know what you think of this. 

Either way, I'm going to need to create all new files and directories, because my steps no longer correspond with those in the `tutorials/darcy_thermo_mech` directory.

**Comment by aeslaughter** [comment]
> We need to consider if we want to shift the focus of this workshop to the user building their own application.

This should be the focus, I would like your steps to finish with git commands, to commit and push the content.

> I foresee an approach where we create an entirely new repo under idaholab account, and make each step a commit, or its own branch, but still require the reader make their own repo and reproduce ours. Let me know what you think of this.

We tried this, it is far too difficult to test and maintain. If you need to make a change to an input then you need to go into each commit and make the change. At least with the folders you can search and replace, plus we also use symlinks when the files do not change.

> Either way, I'm going to need to create all new files and directories, because my steps no longer correspond with those in the `tutorials/darcy_thermo_mech` directory.

Just make a new folder and leave the old one, we can remove the old one when you are done. This will also allow us to start merging content and testing what you have. What name do you have in mind for the folder?

**Comment by crswong888** [comment]
@aeslaughter I see the issue with creating a separate repository. I'll just create my own stork app on the side so that I can be sure my tutorial instructions are 100% reproducible.

That's a good question about the name of the directory. Since we're sort of consolidating all training materials under this new "Training" alias, what if I created a directory like `training/tutorials/darcy_thermo_mech`, and we can put all future tutorials there as well. We could also move the `examples` folder to `training/examples` as well, and I can fix anything that breaks as a result. I don't see a reason to call it anything except `darcy_thermo_mech`, it just needs a root thats different from the `tutorials` one.

I guess I will need to change many things I've done so far to follow the procedure of building a custom application. I will make this my primary focus now.

File: large_media
Comment (crswong888) [reply large_media:1]: This is fine

**Comment by crswong888** [comment]
This pull request was doing some weird things when comparing the diff. Please excuse all the opening and closing and base switching

**Comment by crswong888** [comment]
@aeslaughter I've totally reorganized the commit history here to make it easier for you to review and for anyone else to follow now, or in the future. At each commit, I confirmed that the website builds with no errors. Thus, the HEAD at each commit is a working version of MooseDocs. I stored a backup of the old commit history on my branch [`crswong888:workshop-15232-old`](https://github.com/crswong888/moose/tree/workshop-15232-old). The heads of these two branches are almost identical, but the "old" one has a history that is a lot messier.

I updated the description for this PR, hopefully, for the last time. I will not work develop this PR any further. My goal now is to get this merged. I will branch off and continue working on tutorial content there. Besides addressing your imminent change requests, the only thing that is left to do is to checkout `large_media` at `master`. Currently, all of my media content exists on a branch of `large_media`, which I will merge into `master` once this PR has your approval.

**Comment by crswong888** [comment]
@aeslaughter looks like I broke MASTODON docs because it also has a MooseDown page named `problem_statement.md`. All I'd have to do is add a bit more verbose to the filename, i.e., `!include isoil/problem_statement.md`. 

Although, I'm not sure what the protocol is for fixing a MOOSE application when you break it in the framework. Should I just open a PR in MASTODON and reference this PR?

**Comment by crswong888** [comment]
Alright I made a PR in mastodon to address the issue with the `problem_statement.md` filename: idaholab/mastodon#321.

**Comment by crswong888** [comment]
New branch to continue developing more tutorial content: [`crswong888:tutorial01-15232`](https://github.com/crswong888/moose/tree/tutorial01-15232)

File: python/MooseDocs/extensions/layout.py
Comment (aeslaughter) [inline python/MooseDocs/extensions/layout.py:95]: I am not sold on this command, can you pull this stuff out into another branch/PR? Don't do that until I push up my other changes.

Comment (crswong888) [reply python/MooseDocs/extensions/layout.py:95]: @aeslaughter Yes. Just let me know when you are done.

**Comment by aeslaughter** [comment]
@crswong888 I made quite a few changes, but don't let it be discouraging, what you have done is good and the whole concept is taking shape. This is a huge change and I expect some back and forth on the text as we finish it up.

I have four more files to go through but wanted to get you some of my changes and talk to you about it. **Thanks for all this work**, this is a big and important step to making our docs much better.

**Comment by crswong888** [comment]
After a strong effort to make it work, it became clear that adding `idaholab/babbler` as a submodule is not practical from a maintenance standpoint. I dropped the two commits which handled the Babbler stuff and replaced them with a commit that adds the applications as simple directories in `tutorials/tutorial01_app_development/`.

I'm going to continue to maintain `idaholab/babbler` via my personal repository [`crswong888/babbler_devel`](https://github.com/crswong888/babbler_devel), because I've made this is a very convenient method to handle all of the folders like a Git repo. I can modify the folders, first, in `crswong888/babbler_devel`, then push them as commits to `idaholab/babbler` to verify that everything looks good, and finally use [`babbler_devel/scripts/copy_to_moose.sh`](https://github.com/crswong888/babbler_devel/blob/master/scripts/copy_to_moose.sh) to copy them to `moose/tutorials/tutorial01_app_development/` and add tracking.

When this project nears completion, we must discuss what to do with `idaholab/babbler`. Perhaps we create a shell script to update this repository in `idaholab/moose` that Civet runs whenever changes are made to files matching `moose/tutorials/tutorial01_app_development/*`. This way, we can reference a real, exemplary, repository in the tutorial for users to compare their own work to. Thus, preserving the original intent of creating Babbler.

**Comment by crswong888** [comment]
Moved the scrollbox commit to #15760

**Comment by crswong888** [comment]
@aeslaughter I couldn't resist: I've proposed some significant back-end MooseDocs changes here. Any commit labeled "WIP" is work done in Python.

I am continuing to make changes to actual tutorial content.

**Comment by crswong888** [comment]
@aeslaughter Please see the changes I have pushed when you get the chance.

**Comment by crswong888** [comment]
@aeslaughter okay, we could initiate an update for the build/execution system and move the commit I have here to a separate branch, specifically for this. And then, open an issue for a MooseDocs executioner overhaul. This ticket would then basically just sit on the back burner until the Tutorial is complete, at least.

**Comment by crswong888** [comment]
@aeslaughter

**Review by aeslaughter — CHANGES_REQUESTED** [review]
This is amazing! I will be adding a commit with a few minor changes to the text.

The only thing I want to change is the build.py changes. If you want to improve build.py, then I would like the file timestamp based approach implemented that we talked about. I can help you with the design of it.

**Comment by aeslaughter** [comment]
You will also need to rebase and fix the conflicts. After you do that, could you also add a few sentences to the Nov. newsletter about the content and what else is coming. You won't need to worry about doing that later then.

File: modules/doc/content/getting_started/examples_and_tutorials/tutorial01_app_development/step01_moose_app.md
Comment (aeslaughter) [inline modules/doc/content/getting_started/examples_and_tutorials/tutorial01_app_development/step01_moose_app.md:87]: GitHub now defaults to "main"

**Comment by crswong888** [comment]
@aeslaughter I'm not sure if I'm the best person to compose an official newsletter announcement, but I've at least laid out the basic message.

The most recent commit should not actually be merged. I attempted to fix some broken links with the large_media check out, hoping it would work, which it didn't. But, I didn't want to just throw this away. Ultimately, this commit can be cherry-picked to save some time.

**Review by aeslaughter — DISMISSED** [review]

**Review by aeslaughter — APPROVED** [review]

**Comment by aeslaughter** [comment]
BISON fix is on its way, this created an incompatible docs situation

**Pull request merged**
